### PR TITLE
Adding to StringUtils truncate method and test cases

### PR DESCRIPTION
**Pull request by thiagoh**
Adding the following methods from `StringUtils`:
* `public static String truncate(final String str, final int maxWidth)`
* `public static String truncate(final String str, int offset, final int maxWidth)`

For example 
```
StringUtils.truncate("abcdefg", 4) = "abcd"
StringUtils.truncate("abcdefg", 6) = "abcdef"
StringUtils.truncate("abcdefg", 7) = "abcdefg"
StringUtils.truncate("abcdefg", 8) = "abcdefg"

StringUtils.truncate("abcdefghijklmno", -1, 10) = "abcdefghij"
StringUtils.truncate("abcdefghijklmno", 0, 10) = "abcdefghij"
StringUtils.truncate("abcdefghijklmno", Integer.MIN_VALUE, 10) = "abcdefghij"
StringUtils.truncate("abcdefghijklmno", Integer.MIN_VALUE, Integer.MAX_VALUE) = "abcdefghijklmno"
StringUtils.truncate("abcdefghijklmno", 0, Integer.MAX_VALUE) = "abcdefghijklmno"
StringUtils.truncate("abcdefghijklmno", 1, 10) = "bcdefghijk"
StringUtils.truncate("abcdefghijklmno", 2, 10) = "cdefghijkl"
StringUtils.truncate("abcdefghijklmno", 3, 10) = "defghijklm"
StringUtils.truncate("abcdefghijklmno", 4, 10) = "efghijklmn"
StringUtils.truncate("abcdefghijklmno", 5, 10) = "fghijklmno"
StringUtils.truncate("abcdefghijklmno", 5, 5) = "fghij"
StringUtils.truncate("abcdefghijklmno", 5, 3) = "fgh"
StringUtils.truncate("abcdefghijklmno", 10, 3) = "klm"
StringUtils.truncate("abcdefghijklmno", 10, Integer.MAX_VALUE) = "klmno"
StringUtils.truncate("abcdefghijklmno", 13, 1) = "n"
StringUtils.truncate("abcdefghijklmno", 13, Integer.MAX_VALUE) = "no"
StringUtils.truncate("abcdefghijklmno", 14, 1) = "o"
StringUtils.truncate("abcdefghijklmno", 14, Integer.MAX_VALUE) = "o"
```

@britter I'm rebasing this PR because of my old PR #24. Could you check it? Thanks

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/StringUtils.java [hunk 1]
```diff
@@ -460,6 +460,108 @@ public static String trimToEmpty(final String str) {
...
+     * StringUtils.truncate("abcdefg", 6) = "abcdef"
+     * StringUtils.truncate("abcdefg", 7) = "abcdefg"
+     * StringUtils.truncate("abcdefg", 8) = "abcdefg"
+     * StringUtils.truncate("abcdefg", -1) = ""
+     * </pre>
+     *
+     * @param str  the String to check, may be null
+     * @param maxWidth  maximum length of result String, must be positive
+     * @return truncated String, {@code null} if null String input
+     * @since 4.0
```
Comment (PascalSchumacher) [inline src/main/java/org/apache/commons/lang3/StringUtils.java]: I guess this should be `3.5`?

**Comment by PascalSchumacher** [comment]
JIRA https://issues.apache.org/jira/browse/LANG-1013

**Comment by PascalSchumacher** [comment]
Hi Thiago,

thanks for the pull request. :+1:

As I understand the discussion in the ticket and the old pull request is, a negative `offset` and/or a negative `maxWidth` should cause an `IllegalArgumentException` (like the javadoc says).

It would be nice if you could update the pull request.

Thanks,
Pascal

File: src/main/java/org/apache/commons/lang3/StringUtils.java [hunk 2]
```diff
@@ -460,6 +460,108 @@ public static String trimToEmpty(final String str) {
...
+     * StringUtils.truncate("abcdefghijklmno", 15, Integer.MAX_VALUE) = ""
+     * StringUtils.truncate("abcdefghijklmno", Integer.MAX_VALUE, Integer.MAX_VALUE) = ""
+     * StringUtils.truncate("abcdefghij", 0, -1) = ""
+     * </pre>
+     *
+     * @param str  the String to check, may be null
+     * @param offset  left edge of source String
+     * @param maxWidth  maximum length of result String, must be positive
+     * @return truncated String, {@code null} if null String input
+     * @since 4.0
```
Comment (PascalSchumacher) [inline src/main/java/org/apache/commons/lang3/StringUtils.java]: should be `3.5`

**Comment by thiagoh** [comment]
@PascalSchumacher sure! I'll fix it

**Comment by thiagoh** [comment]
@PascalSchumacher can you check this out?

File: src/main/java/org/apache/commons/lang3/StringUtils.java [hunk 3]
```diff
@@ -460,6 +460,118 @@ public static String trimToEmpty(final String str) {
...
+     * </ul>
+     *
+     * <pre>
+     * StringUtils.truncate(null, *)      = null
+     * StringUtils.truncate("", 4)        = ""
+     * StringUtils.truncate("abcdefg", 4) = "abcd"
+     * StringUtils.truncate("abcdefg", 6) = "abcdef"
+     * StringUtils.truncate("abcdefg", 7) = "abcdefg"
+     * StringUtils.truncate("abcdefg", 8) = "abcdefg"
+     * StringUtils.truncate("abcdefg", -1) = ""
```
Comment (PascalSchumacher) [inline src/main/java/org/apache/commons/lang3/StringUtils.java]: This now causes an `IllegalArgumentException`.

**Comment by PascalSchumacher** [comment]
Thanks for updating the pull request. It looks good. :)

Just two small issues:

*  javadoc examples have to be updated for the `IllegalArgumentException` for negative offset and length
* checking offset and length should be done at the very start of the method (e.g. do not allow a negative offset and/or length for a `null` string)

Thanks,
Pascal

**Comment by thiagoh** [comment]
@PascalSchumacher I think now thats. Could you confirm that?

**Comment by PascalSchumacher** [comment]
Thanks! :+1:

**Pull request closed without merging**